### PR TITLE
Clean interface for public instances between React and React Native

### PR DESCRIPTION
**Pull request by rubennorte**
## Summary

We are going to move the definition of public instances from React to React Native to have them together with the native methods in Fabric that they invoke. This will allow us to have a better type safety between them and iterate faster on the implementation of this proposal: https://github.com/react-native-community/discussions-and-proposals/pull/607

The interface between React and React Native would look like this after this change and a following PR (#26418):

React → React Native:
```javascript
ReactNativePrivateInterface.createPublicInstance // to provide via refs
ReactNativePrivateInterface.getNodeFromPublicInstance // for DevTools, commands, etc.
ReactNativePrivateInterface.getNativeTagFromPublicInstance // to implement `findNodeHandle`
```

React Native → React (ReactFabric):
```javascript
ReactFabric.getNodeFromInternalInstanceHandle // to get most recent node to call into native
ReactFabric.getPublicInstanceFromInternalInstanceHandle // to get public instances from results from native
```

## How did you test this change?

Flow
Existing unit tests

### CODE REVIEW

File: packages/react-native-renderer/src/ReactNativeFiberInspector.js [hunk 1]
```diff
@@ -196,12 +196,7 @@ function getInspectorDataForViewAtPoint(
   if (__DEV__) {
     let closestInstance = null;
 
-    const fabricInstanceHandle =
```
Comment (rubennorte) [inline packages/react-native-renderer/src/ReactNativeFiberInspector.js:199]: This level of indirection was unnecessary.

File: packages/react-native-renderer/src/ReactFabric.js [hunk 2]
```diff
@@ -124,6 +125,10 @@ export {
   // This method allows it to acess the most recent shadow node for
   // the instance (it's only accessible through it).
   getNodeFromInternalInstanceHandle,
+  // Fabric native methods to traverse the host tree return the same internal
+  // instance handles we use to dispatch events. This provides a way to access
+  // the public instances we created from them (potentially created lazily).
+  getPublicInstanceFromInternalInstanceHandle,
```
Comment (rickhanlonii) [inline packages/react-native-renderer/src/ReactFabric.js:131]: What's the difference between a Node and a PublicInstance (i.e. why doesn't the existing `getNodeFromInternalInstanceHandle` work)?

Comment (rubennorte) [reply packages/react-native-renderer/src/ReactFabric.js:131]: Node is the shadow node (which is internal). Public instance is the instance of `ReactFabricHostComponent` that's provided via refs.

Comment (rickhanlonii) [reply packages/react-native-renderer/src/ReactFabric.js:131]: I wonder if these are supposed to be different things? I think I would expect a Node to be a HostComponent but I don't have a ton of context.

**Review by sammy-SC — APPROVED** [review]

**Pull request merged**